src/base/LruMap.h

```diff
@@ -148,39 +169,47 @@ LruMap<EntryValue, EntryCost>::add(const char *key, EntryValue *t)
     if (memLimit() == 0)
         return false;
 
-    index.push_front(new Entry(key, t));
+    auto e = new Entry(key, t, ttl);
+    memUsed_ += e->memCost() + e->key.length() + sizeof(Entry) + sizeof(MapPair);
+    index.push_front(e);
     storage.insert(MapPair(key, index.begin()));
 
     ++entries_;
+
+    // if over memory limit, start purging
+    while (memUsed_ > memLimit_)
+        del(storage.end());
+
     return true;
 }
 
-template <class EntryValue, size_t EntryCost>
+template <class EntryValue, size_t EntryCost(const EntryValue *)>
 bool
 LruMap<EntryValue, EntryCost>::expired(const LruMap::Entry &entry) const
 {
-    if (ttl < 0)
+    if (retention < 0)
         return false;
 
-    return (entry.date + ttl < squid_curtime);
+    return (entry.expires < squid_curtime);
 }
 
-template <class EntryValue, size_t EntryCost>
+template <class EntryValue, size_t EntryCost(const EntryValue *)>
 bool
 LruMap<EntryValue, EntryCost>::del(LruMap::MapIterator const &i)
 {
     if (i != storage.end()) {
         Entry *e = *i->second;
         index.erase(i->second);
         storage.erase(i);
+        memUsed_ -= (e->memCost() + e->key.length() + sizeof(Entry) + sizeof(MapPair));
```

Comment:
I have slightly reorganized these memory accounting lines. e->memCost() should now account for all the Entry memory usage. Added a memoryUsedBy() method for the MapPair objects.